onnx/defs/math/defs.cc

```diff
@@ -2006,6 +2007,301 @@ ONNX_OPERATOR_SET_SCHEMA(
+          if (block_size < 16 || ((block_size & (block_size - 1)) != 0)) {
+            fail_type_inference("block_size attribute must be a a power of 2 and not smaller than 16.");
+          }
+
+          propagateElemTypeFromInputToOutput(ctx, 0, 0);
+
+          auto a_type = ctx.getInputType(0);
+          auto b_type = ctx.getInputType(1);
+          auto scales_type = ctx.getInputType(2);
+          auto y_type = ctx.getOutputType(0);
+          if (nullptr == a_type || nullptr == b_type || nullptr == scales_type || nullptr == y_type ||
+              a_type->value_case() != ONNX_NAMESPACE::TypeProto::kTensorType ||
+              b_type->value_case() != ONNX_NAMESPACE::TypeProto::kTensorType ||
+              scales_type->value_case() != ONNX_NAMESPACE::TypeProto::kTensorType) {
+            fail_type_inference("inputs are expected to have tensor type and output type should not be null.");
+          }
+
+          if (!hasNInputShapes(ctx, 3)) { return; }
+
+          const auto a_shape = a_type->tensor_type().shape();
+          const auto b_shape = b_type->tensor_type().shape();
+          const auto scales_shape = scales_type->tensor_type().shape();
+          if (a_shape.dim_size() != 2) {
+            fail_type_inference("Input A tensor of wrong rank.");
+          }
+          if (b_shape.dim_size() < 2 || b_shape.dim_size() > 3) {
+            fail_type_inference("Input B tensor of wrong rank.");
+          }
+          if (scales_shape.dim_size() != 1) {
+            fail_type_inference("Input scales tensor of wrong rank.");
+          }
+
+          // if K attribute is not present, it should be inferred from input A
+          if ( K == -1 && a_shape.dim(1).has_dim_value()) {
+            K = a_shape.dim(1).dim_value();
+          }
+
+          // if N attribute is not present, it should be inferred from input B
+          if ( N == -1 && b_shape.dim(0).has_dim_value()) {
+            N = b_shape.dim(0).dim_value();
+          }
+
+          // Input B is stored as uint8_t with shape: `[N][n_blocks_per_col][blob_size]`
+          // in which:
+          //  - `n_blocks_per_col` = `(K + block_size - 1) / block_size`
+          //  - `blob_size` = `CeilDiv(block_size * bits, 8)`
+          // intput shape A{M, K} B{N, n_blocks_per_col, blob_size}
+          if (a_shape.dim(1).has_dim_value() && a_shape.dim(1).dim_value() != K) {
+            fail_shape_inference("Incompatible dimensions for matrix multiplication. "
+                                 "Shape A and K attribute don't match.");
+          }
+
+          if (b_shape.dim(0).has_dim_value() && b_shape.dim(0).dim_value() != N) {
+            fail_shape_inference("Incompatible dimensions for matrix multiplication. "
+                                 "Shape B and N attribute don't match.");
+          }
+
+          int64_t n_blocks_per_col = (K + block_size - 1) / block_size;
```

Comment:
In rare cases, K could be -1 here (unknown). We should skip corresponding checks involving n_blocks_per_col in that case.